62.CAD/main.cpp

```diff
@@ -3,31 +3,177 @@
 
 #include "../common/CommonAPI.h"
 #include "nbl/ext/FullScreenTriangle/FullScreenTriangle.h"
+#include "nbl/core/SRange.h"
 #include "glm/glm/glm.hpp"
 #include <nbl/builtin/hlsl/cpp_compat/matrix.hlsl>
 #include <nbl/builtin/hlsl/cpp_compat/vector.hlsl>
 #include "curves.h"
 
-static constexpr bool DebugMode = false;
-static constexpr bool FragmentShaderPixelInterlock = true;
+#include <nbl/builtin/hlsl/shapes/beziers.hlsl>
 
 enum class ExampleMode
 {
 	CASE_0, // Simple Line, Camera Zoom In/Out
 	CASE_1,	// Overdraw Fragment Shader Stress Test
 	CASE_2, // NOT USED
 	CASE_3, // CURVES AND LINES
+	CASE_4, // STIPPLE PATTERN
 };
 
-constexpr ExampleMode mode = ExampleMode::CASE_3;
-using namespace nbl::hlsl;
+constexpr ExampleMode mode = ExampleMode::CASE_4;
+static constexpr bool DebugMode = false;
+static constexpr bool FragmentShaderPixelInterlock = false;
 
 #include "common.hlsl"
 
+using namespace nbl::hlsl;
+
+bool operator==(const LineStyle& lhs, const LineStyle& rhs)
+{
+	const bool areParametersEqual =
+		lhs.color == rhs.color &&
+		lhs.screenSpaceLineWidth == rhs.screenSpaceLineWidth &&
+		lhs.worldSpaceLineWidth == rhs.worldSpaceLineWidth &&
+		lhs.stipplePatternSize == rhs.stipplePatternSize &&
+		lhs.recpiprocalStipplePatternLen == rhs.recpiprocalStipplePatternLen &&
+		lhs.phaseShift == rhs.phaseShift;
+
+	if (!areParametersEqual)
+		return false;
+
+	if (lhs.stipplePatternSize == -1)
+		return true;
+
+	assert(lhs.stipplePatternSize <= LineStyle::STIPPLE_PATTERN_MAX_SZ && lhs.stipplePatternSize >= 0);
+
+	const bool isStipplePatternArrayEqual = std::memcmp(lhs.stipplePattern, rhs.stipplePattern, sizeof(decltype(lhs.stipplePatternSize)) * lhs.stipplePatternSize);
+
+	return areParametersEqual && isStipplePatternArrayEqual;
+}
+
+// holds values for `LineStyle` struct and caculates stipple pattern re values, cant think of better name
+class CPULineStyle
+{
+public:
+	static const uint32_t STIPPLE_PATTERN_MAX_SZ = 15u;
+
+	// common data
+		// private and setters/getters instead?
```

Comment:
nah, and make this a struct and get rid of `public`

Reply:
done